From d7e3ac16d389bfd9dd2a394e0bddefd32dd50452 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 16 Oct 2014 18:23:14 -0400
Subject: [ethernet] cxgb4vf: Call dev_kfree/consume_skb_any instead of [dev_]kfree_skb

Message-id: <1413483794-20865-1-git-send-email-svemuri@redhat.com>
Patchwork-id: 97616
O-Subject: [RHEL7.1 PATCH BZ 1070945 v2] cxgb4vf: Call dev_kfree/consume_skb_any instead of [dev_]kfree_skb.
Bugzilla: 1070945

Replace kfree_skb with consume_skb in free_tx_desc that can be
called in hard irq and other contexts. consume_skb is used
as this function consumes successfully transmitted skbs.

Replace dev_kfree_skb with dev_kfree_skb_any in t4vf_eth_xmit that can
be called in hard irq and other contexts, on paths that drop the skb.

Replace dev_kfree_skb with consume_skb in t4vf_eth_xmit that can
be called in hard irq and other contexts, on paths that successfully
transmit the skb.

Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
(cherry picked from commit 42ffda5fe704ab6a3ef8ca36b5f958a3cd86a164)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4vf/sge.c b/drivers/net/ethernet/chelsio/cxgb4vf/sge.c
index a48ca91..2360546 100644
--- a/drivers/net/ethernet/chelsio/cxgb4vf/sge.c
+++ b/drivers/net/ethernet/chelsio/cxgb4vf/sge.c
@@ -401,7 +401,7 @@ static void free_tx_desc(struct adapter *adapter, struct sge_txq *tq,
   if (sdesc->skb) {
    if (need_unmap)
     unmap_sgl(dev, sdesc->skb, sdesc->sgl, tq);
-   kfree_skb(sdesc->skb);
+   consume_skb(sdesc->skb);
    sdesc->skb = NULL;
   }
 
@@ -1275,7 +1275,7 @@ int t4vf_eth_xmit(struct sk_buff *skb, struct net_device *dev)
    * need it any longer.
    */
   inline_tx_skb(skb, &txq->q, cpl + 1);
-  dev_kfree_skb(skb);
+  consume_skb(skb);
  } else {
   /*
    * Write the skb's Scatter/Gather list into the TX Packet CPL
@@ -1354,7 +1354,7 @@ out_free:
   * An error of some sort happened.  Free the TX skb and tell the
   * OS that we've "dealt" with the packet ...
   */
- dev_kfree_skb(skb);
+ dev_kfree_skb_any(skb);
  return NETDEV_TX_OK;
 }
 
-- 
1.7.1